### ISSUE TITLE
Autodetection of properly tagged dynamic formats is non-working

### ISSUE THREAD

**Issue by solardiz** [openwall/john#4002]
As mentioned in:

https://github.com/magnumripper/JohnTheRipper/issues/3970#issuecomment-493723879

autodetection of properly tagged dynamic formats is currently non-working. I think it used to work before. We need to repair it.

The above example is for dynamic_82, but I just ran into the same issue with dynamic_62.

**Comment by jfoug** [comment]
This is most certainly by design (not mine).   The file ./run/dynamic_disabled.conf lists a set of dynamic hashes which have had their auto-* logic removed. The will not test, run or do anything, UNLESS you add the -format= tag.

Both dynamic_82 and dynamic_62 are in that file.

In john.conf : 

```
# These formats are disabled from all-formats --test runs, or auto-selection
# of format from an input file. Even when disabled, you can use them as long
# as you spell them out with the --format option. Or you can delete a line,
# comment it out, or change to 'N'
[Disabled:Formats]
#formatname = Y
.include '$JOHN/dynamic_disabled.conf'
```

then dynamic_disabled.conf is simply a long list of dynamic hash types.  All of them are disabled.

**Comment by magnumripper** [comment]
You added it to `john.conf` in 3aa49af9, Jim - and then moved them to its own file in facddf0b 😉 

I was probably involved in some discussion leading to that though. I think we meant to only put really academic formats in there so perhaps dynamic_82 (and dynamic_62?) should be dropped from that list.

**Comment by jfoug** [comment]
are 82 and 62 the only ones?

**Comment by magnumripper** [comment]
I have no idea. Ideal would be a (very) carefully thought out reduced self-test that only tries each "code path" of dynamic. For example, `md5($p)`, `sha1($s.$p)`, `keccak($p.$s)` and so on. But IIRC what we wanted with this was to not (by default) test things like `ripemd160(ripemd160($p).ripemd160($p))` which was likely never used anywhere IRL and the reason was the sheer number of dynamics.

**Comment by jfoug** [comment]
So in ideal, we would have all dynamic formats work 100% of the time EXCEPT for -test* switches (unless -format was added).  Is that the idea?

I think the way it was handled was that the disabled simply did not load the format, unless it was specified.   This certainly keeps -test* from blowing through all the dynamic academic types.  But we DO want john to run properly on input files for any dynamic type (academic or not).

If that is the idea, then I can start looking into it.

**Comment by jfoug** [comment]
I have some more testing to do, but this may be easy.  I simply moved this code (and a bit that sets override_disable) from john.c (register_all_formats) into bench.c  ( benchmark_all).   I think that should be all that is needed to make the disabled simply NOT test (unless forced), but still should load the formats, and THUS, should recognize hashes in an input file.

```c
	/* Format disabled in john.conf */
	if (!override_disable &&
	    cfg_get_bool(SECTION_DISABLED, SUBSECTION_FORMATS,
	                 format->params.label, 0)) {
#ifdef DEBUG
		if (format->params.flags & FMT_DYNAMIC) {
			/* in debug mode, we 'allow' dyna */
		} else
#else
		if (options.format &&
		    !strcasecmp(options.format, "dynamic-all") &&
		    (format->params.flags & FMT_DYNAMIC)) {
			wildcard_format = 1;
			/* allow dyna if '-format=dynamic-all' was selected */
		} else
#endif
		if (options.format &&
		    !strcasecmp(options.format, format->params.label)) {
			/* allow if specifically requested */
		} else
			return;
	}
```

**Comment by jfoug** [comment]
PR #4324 should resolve this issue.  With that fix, all formats will properly auto-detect (even formats in the disabled conf file).

**Comment by solardiz** [comment]
Assumed fixed by #4324
